### PR TITLE
SoftwareSerialM for SKR STM32F1 boards

### PR DESCRIPTION
**Pull request by sjasonsmith**
### Description

This provides two improvements:

1. For SKR Mini / E3 DIP / E3 Mini users, provides a working default SoftwareSerial implementation
    - Uses Timer 4, which is the same timer BigTreeTech/TMCStepper used

1. For other STM32F1 users:
    - I did not enable this there, because effort is needed to select a safe timer to use
    - Fail build with meaningful message if SoftwareSerial is needed but missing

Output when using TMC220x drivers without HAVE_SW_SERIAL defined:

```
Marlin\src\HAL\HAL_STM32F1\SoftwareSerial.cpp:23:4: warning: #warning "Consider using SoftwareSerialM with HAVE_SW_SERIAL and appropriate SS_TIMER defined. [-Wcpp]
   #warning "Consider using SoftwareSerialM with HAVE_SW_SERIAL and appropriate SS_TIMER defined.
    ^~~~~~~
Marlin\src\HAL\HAL_STM32F1\SoftwareSerial.cpp:24:4: error: #error "Missing SoftwareSerial implementation"
   #error "Missing SoftwareSerial implementation"
    ^~~~~
```

### Benefits

SoftwareSerial will just work for anyone using these boards, and users of other STM32F1 boards won't waste time debugging to learn they are using the dummy SoftwareSerial!

Hopefully others will contribute similar changes for the STM32F1 boards they use.

I have also verified that building without SoftwareSerial works fine when using 2130 drivers. In this case the default dummy implementation is used to satisfy TMCStepper build issues.

### Related Issues

Somehow I cannot find a single open issue about this, although it seems to be a popular topic on Discord.

### CODE REVIEW

File: Marlin/src/HAL/HAL_STM32F1/SoftwareSerial.cpp [hunk 1]
```diff
@@ -18,6 +18,12 @@
  */
 #if defined(__STM32F1__) && !defined(HAVE_SW_SERIAL)
 
+#include "../../inc/MarlinConfig.h"
+#if HAS_TMC220x
+  #warning "Consider using SoftwareSerialM with HAVE_SW_SERIAL and appropriate SS_TIMER defined.
```
Comment (anlupat) [inline Marlin/src/HAL/HAL_STM32F1/SoftwareSerial.cpp]: missing the ending "

Comment (sjasonsmith) [reply Marlin/src/HAL/HAL_STM32F1/SoftwareSerial.cpp]: Thanks. I thought I had already pushed that fix while posting the PR. I’ll get it updated shortly.

**Comment by sjasonsmith** [comment]
I undid most of my changes in SoftwareSerial.h. I realized that compiling successfully relied on the include order happening to hit the external library before hitting our HAL.

**Comment by swilkens** [comment]
I tested this on my SKR Mini E3 V1.2 by pulling the most recent bugfix-2.0.x and cherry picking this PR. Everything worked perfectly, running a test print as we speak. Excellent work.

----

p.s.

Does this provide, partially, the same as PR#15655?

https://github.com/MarlinFirmware/Marlin/pull/15655

It seems that STM5.7.0 also provides a softwareserial, do we still need the FYSETC implementation?

**Comment by sjasonsmith** [comment]
> Does this provide, partially, the same as PR#15655?

For context, I also contributed the SoftwareSerial changes in that PR, just indirectly through @LinoBarreca's branch.

These only overlap for the SKR Pro. If #15655 is merged first, then I will update this PR to remove the SKR Pro change. I had earlier pointed SKR Pro to my own fork which had SKR Pro fixes. Those are now merged into the FYSETC repo, so I'm getting rid of my diverging branch from the platformio file. SoftwareSerialM is no longer needed at all for the SKR Pro once #15655 is merged.

> It seems that STM5.7.0 also provides a softwareserial, do we still need the FYSETC implementation?

SoftwareSerial doesn't come directly from STM32 5.7, it comes from Arduino_Core_STM32. This is currently only used by HAL_STM32, and not HAL_STM32F1. Perhaps HAL_STM32 will be used for all STM32 boards in the future, but for now STM32F1 boards still need an external SoftwareSerial.

For the SKR Pro #15655 includes a copy of SoftwareSerial directly in HAL_STM32, which fixes some bugs in the Arduino_Core_STM32 version. This is temporary while we wait for an Arduino_Core_STM32 1.8 release, where we have already fixed those issues.

**Comment by swilkens** [comment]
So for future reference - as I'm sure many will refer to this PR over the next weeks - for the time being (also after #15655 is merged) - HAL_STM32F1 boards like the SKR Mini E3 will still need to use the external FYSETC SoftwareSerial provided by this PR.

Those boards can then also use upstream TMCStepper, rather than the BTT fork.

Great work, thanks!

**Comment by sjasonsmith** [comment]
> HAL_STM32F1 boards like the SKR Mini E3 will still need to use the external FYSETC SoftwareSerial provided by this PR.

Unless you have an SKR Mini E3 1.0, which uses HardwareSerial! I was really disappointed that 1.2 switched to SoftwareSerial, presumably so they could downgrade to TMC2208 drivers if they need to.

**Comment by LinoBarreca** [comment]
@sjasonsmith are you on fb?

**Comment by damvcoool** [comment]
> Unless you have an SKR Mini E3 1.0, which uses HardwareSerial! I was really disappointed that 1.2 switched to SoftwareSerial, presumably so they could downgrade to TMC2208 drivers if they need to.

I have a SKR Mini E3 v1.2 (not DIP) and it has TMC2209.

if the HW serial is part of the STM32F1, and the SKR Mini E3 1.0 has STM32F103RCT6 and v1.2 also has STM32F103RCT6, wouldn't both have a HW serial? or it depends on how the TMC2209 drivers were wired?

**Comment by sjasonsmith** [comment]
> if the HW serial is part of the STM32F1, and the SKR Mini E3 1.0 has STM32F103RCT6 and v1.2 also has STM32F103RCT6, wouldn't both have a HW serial? or it depends on how the TMC2209 drivers were wired?

Four 2209 drivers can share the same serial pins, so they can all share one hardware serial port. Each 2208 needs separate pins, so software serial becomes necessary.

In all likelihood they changed this to give them more flexibility to switch drivers if TMC2209s end up too hard to obtain.

**Comment by sjasonsmith** [comment]
I've merged the recent timing changes into this. The review has been updated with the following changes:

1. Revert the change for SKR Pro, since SoftwareSerial is now included in the HAL
1. Revert changes to STM32F103RE environment. I had only intended to leave the bigtree variants boards modified. I would prefer that other environments be changed by someone who can actually verify they work with boards from other vendors.

**Comment by thinkyhead** [comment]
> I realized that compiling successfully relied on the include order happening to hit the external library before hitting our HAL.

Let's be sure to use `#include <filename.h>` rather than `#include "filename.h"` for all libraries and platform headers.

**Comment by sjasonsmith** [comment]
> Let's be sure to use `#include <filename.h>` rather than `#include "filename.h"` for all libraries and platform headers.

@thinkyhead In this case I never saw any build failure, but since this is included from other libraries (TMCStepper) I didn't want to rely on their includes encountering the library version first. Perhaps that order would be guaranteed, but I wasn't sure if that would be dependable over time.

**Comment by thinkyhead** [comment]
We can guarantee include order of `.h` files where it matters. Just point out anyplace where include order is problematic, and we can address it by reordering includes or renaming files.

**Comment by sjasonsmith** [comment]
No current problems that I'm aware of.

**Comment by hobiseven** [comment]
@sjasonsmith 
Quick one : browsing through some SKR pins files, it seems that both RX and TX pins can be the same in this SW UART implementation.... 
Is this correct?

**Comment by sjasonsmith** [comment]
> @sjasonsmith
> Quick one : browsing through some SKR pins files, it seems that both RX and TX pins can be the same in this SW UART implementation....
> Is this correct?

@hobiseven, yes, that is correct. That is also true for LPC1768/9 and AVR boards as well.

**Pull request merged**